### PR TITLE
feat: Create a default Dockerfile

### PR DESCRIPTION
**Pull request by helio-frota**

### CODE REVIEW

**Comment by helio-frota** [comment]
Hi @lholmquist this PR is a WIP, and some situations still need to be solved:

* We need to guarantee that if we create a Dockerfile, it will not end up on the git repository of the user, because it can looks weird with a remaining new file untracked, so we need to remove the Dockerfile created after all the processing.

* if the project already has a Dockerfile how we can avoid this Dockerfile be deleted ?

* We can try to use [tar-stream](https://www.npmjs.com/package/tar-stream) to avoid Dockerfile creation on disk, and then not bother with checking if the user already wrote a Dockerfile

File: lib/dockerfile-writer.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+'use strict';
+
+// This module will be used to create a Dockerfile in project root directory when using Docker build strategy.
+
+const { promisify } = require('util');
+const fs = require('fs');
+const writeFileAsync = promisify(fs.writeFile);
+const logger = require('./common-log')();
+
+async function writeDockerfile (config, from='node:14', app='index.js') {
```
Comment (helio-frota) [inline lib/dockerfile-writer.js:28]: @lholmquist 
* Another thing that is not clear yet is about the arguments `from, app` it is possible for config object to hold these arguments ?

File: lib/goals/build.js [hunk 2]
```diff
@@ -25,11 +25,15 @@ const projectArchiver = require('../project-archiver');
...
 const imageStreamConfigurator = require('../image-stream');
 const binaryBuild = require('../binary-build');
+const helpers = require('../helpers');
 
 module.exports = async function build (config) {
   // archive the application source
   await projectArchiver.archiveAndTar(config);
 
+  // Removes the Dockerfile created by nodeshift.
+  await helpers.removeDockerfile(config.projectLocation);
```
Comment (helio-frota) [inline lib/goals/build.js:35]: I'm not comfortable with this approach of removing the Dockerfile here in this location but it needs to happen after the tar creation if we decide to create the file on the disk. Or we can add this a documentation for the user that he will end up with a new file in their git repository after that (?)

**Comment by lholmquist** [comment]
:eyes:

**Comment by lholmquist** [comment]
My initial thought,  was that if they are using the docker build strategy and they do not have a dockerfile in their project directory,  then we warn them that we are going to create a default one.

I think that default Dockerfile could be created in the `tmp/` directory where we store the archive.tar file.  then maybe we just use the `[tar.replace](https://www.npmjs.com/package/tar#tarroptions-filelist-callback-alias-tarreplace)` function to add that file to the archive we created, so we don't need to create anything in the users project.   

that `tmp/` directory gets deleted during every call to nodeshift during the build phase.

**Comment by helio-frota** [comment]
Sounds great to use `/tmp` directory, I'll make more tests locally and see what happen, thanks

**Comment by lholmquist** [comment]
> when I use confg.build.strategy the tests fails. 

You will need to add an empty build object to the config in the failing tests.  Normally, nodeshift will add that in the nodeshift-config code,  but since we are mocking,  we need to add that

**Comment by lholmquist** [comment]
> Probably this from can be replaced to config.build.stategy.from or something like that according #447 (?)

Probably,  but lets leave that for another PR

**Comment by helio-frota** [comment]
@lholmquist I'm going to close this draft and revisit later as discussed other day :+1:

**Pull request closed without merging**